File: src/main/java/no/nav/sosialhjelp/soknad/business/service/dialog/SistInnsendteSoknadService.java
package no.nav.sosialhjelp.soknad.business.service.dialog;

import no.nav.sosialhjelp.soknad.business.db.repositories.soknadmetadata.SoknadMetadataRepository;
import no.nav.sosialhjelp.soknad.web.rest.ressurser.dialog.dto.SistInnsendteSoknadDto;
import org.springframework.stereotype.Component;

import java.util.Optional;

import static java.util.Collections.emptyList;
import static java.util.Comparator.comparing;

@Component
public class SistInnsendteSoknadService {

    private final SoknadMetadataRepository soknadMetadataRepository;

    public SistInnsendteSoknadService(
            SoknadMetadataRepository soknadMetadataRepository
    ) {
        this.soknadMetadataRepository = soknadMetadataRepository;
    }

    public Optional<SistInnsendteSoknadDto> hentSistInnsendteSoknad(String fnr) {
        return Optional.ofNullable(soknadMetadataRepository.hentAlleInnsendteSoknaderForBruker(fnr)).orElse(emptyList())
                .stream()
                .max(comparing(o -> o.innsendtDato))
                .map(metadata -> new SistInnsendteSoknadDto(metadata.fiksForsendelseId, metadata.navEnhet, metadata.innsendtDato));
    }
}

File: src/main/kotlin/no/nav/sosialhjelp/soknad/api/dialog/SistInnsendteSoknadService.kt
package no.nav.sosialhjelp.soknad.api.dialog

import no.nav.sosialhjelp.soknad.api.dialog.dto.SistInnsendteSoknadDto
import no.nav.sosialhjelp.soknad.db.repositories.soknadmetadata.SoknadMetadataRepository
import org.springframework.stereotype.Component
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

@Component
class SistInnsendteSoknadService(
    private val soknadMetadataRepository: SoknadMetadataRepository
) {
    fun hentSistInnsendteSoknad(fnr: String): SistInnsendteSoknadDto? {
        val sistInnsendteSoknad = soknadMetadataRepository.hentAlleInnsendteSoknaderForBruker(fnr)
            .maxByOrNull { it.innsendtDato ?: LocalDateTime.MIN }
        val navEnhet = sistInnsendteSoknad?.navEnhet ?: return null
        val innsendtDato = sistInnsendteSoknad.innsendtDato ?: return null
        return SistInnsendteSoknadDto(
            sistInnsendteSoknad.fnr,
            navEnhet,
            innsendtDato.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME)
        )
    }
}
